### ISSUE TITLE
CNN w variable sized input performance regression 1.10.2 cu113 -> 2.0.1 cu117

### ISSUE THREAD

**Issue by jxtps** [pytorch/pytorch#108474]
### 🐛 Describe the bug

I'm running an image processing service that takes images of varying sizes and runs them through a couple of CNNs, one with fixed size input (always w0 * h0) and one with variable size input (w & h vary from input to input). 

I've been trying to upgrade from v1.10.2 cu113 to v2.0.1 cu117 and encountered https://github.com/bytedeco/javacpp-presets/issues/1409 / https://github.com/pytorch/pytorch/pull/104369. When I apply the workaround (`TORCH_CUDNN_V8_API_DISABLED=1`) and run a load test I see the following performance regression: 

![image](https://github.com/pytorch/pytorch/assets/90846/f4a2eefe-ed81-4a7e-a5c3-278e473af6ca)

In addition, the first couple of runs through my networks can easily take ~30 seconds on v2.0.1 when they take <1 second on v1.10.2. 

This performance regression makes it hard to upgrade and I'm not really sure how to proceed here - I have an entire production edifice where it exhibits, but it's all proprietary so hard to share in a way that could enable debugging. 

I guess I'd need to find a way to replicate it in a smaller python-only example? 

### Versions

The issue exhibits in v2.0.1 cu117 but not in v1.10.2 cu113

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by jxtps** [comment]
Turns out this regression was already present in 1.12.1 (though it looks even worse in 2.0.1): 
![image](https://github.com/pytorch/pytorch/assets/90846/776df08e-363e-442f-ad76-fa6efa7b4c9b)

It also has the ~30s "freeze" on first call to some networks. 

* Ubuntu 20.04.6 LTS (GNU/Linux 5.15.0-1040-aws x86_64)
* AWS g5.2xlarge instance, A10g NVIDIA processor

I'll try to put together a python-only reproduction.

**Comment by jxtps** [comment]
This little script seems to show enough of an issue, at least on my Windows 10 machine with a 3090: 
```python
import math
import os
import time
from random import Random

os.environ['PYTORCH_CUDA_ALLOC_CONF'] = 'max_split_size_mb:512'
os.environ['TORCH_CUDNN_V8_API_DISABLED'] = '1'

import torch

class TestModel(torch.nn.Sequential):
    def __init__(self, in_chans=4, ch64: int = 64):
        super().__init__(
            torch.nn.Conv2d(in_chans, ch64, 3, padding=1),
            torch.nn.ReLU(inplace=True),
            torch.nn.Conv2d(ch64, ch64, 3, padding=1),
            torch.nn.ReLU(inplace=True),
            torch.nn.Conv2d(ch64, in_chans, 3, padding=1),
        )

if __name__ == '__main__':
    print(f'MPx {torch.__version__}\tGpuMs {torch.__version__}')
    use_channels_last = False  # <-- Change this to True if you have a fancy GPU, e.g. A100, A10g, etc
    mf = torch.channels_last if use_channels_last else torch.contiguous_format
    net = TestModel().to(device='cuda', memory_format=mf)
    rnd = Random(42)
    w = h = 512
    while True:
        px = (rnd.random() * 3.9 + 0.1) * 1e6
        aspect_ratio = rnd.random() + 0.5
        w = int(math.sqrt(px * aspect_ratio))
        h = int(px / w)
        x = torch.empty((1, 4, w, h), device='cuda', memory_format=mf)
        t0 = time.time()
        y = net(x).to(device='cpu')  # Transfer to CPU ensures synchronized & is what the end-user experiences = the relevant metric
        dt = time.time() - t0
        print(f'{w * h * 1e-6:.3f}\t{dt*1000:.1f}')
```
When I run this and let it collect 119 samples for each of 1.10.0 (sorry, it was the version I had on my dev machine), 1.12.1, and 2.0.1 I get the following graphs (note that they all ran *precisely the same resolutions* since `Random(42)` was initialized the same way): 

![image](https://github.com/pytorch/pytorch/assets/90846/8ae08c49-7f3f-446e-9749-738f2ee2d42d)

The averages tell an even clearer story: 

* 1.10.0: 416ms
* 1.12.1: 992ms
* 2.0.1: 838ms

Basically there are seem to just be a lot more outliers with the more recent versions, with 2.0.1 regaining some of the ground lost in 1.12.1. 

The astute observer will of course note the difference in behavior <1MPx compared to the first set of charts. That's presumably either due to the change in GPU (A10g -> 3090) or network. To be continued...

**Comment by jxtps** [comment]
Here's another version of the repro script, switching to `float16`, using a fixed number of iterations, switching to `torch.cuda.synchronize`, and improving the headers: 
```python
import math
import os
import time
from random import Random

from models.network_u2net import U2NET

os.environ['PYTORCH_CUDA_ALLOC_CONF'] = 'max_split_size_mb:512'
os.environ['TORCH_CUDNN_V8_API_DISABLED'] = '1'

import torch


class TestModel(torch.nn.Sequential):
    def __init__(self, in_chans=4, ch64: int = 64):
        super().__init__(
            torch.nn.Conv2d(in_chans, ch64, 3, padding=1),
            torch.nn.ReLU(inplace=True),
            torch.nn.Conv2d(ch64, ch64, 3, padding=1),
            torch.nn.ReLU(inplace=True),
            torch.nn.Conv2d(ch64, in_chans, 3, padding=1),
        )


if __name__ == '__main__':
    dtype = torch.float16
    use_channels_last = False
    mf = torch.channels_last if use_channels_last else torch.contiguous_format
    net = TestModel().to(dtype=dtype, device='cuda', memory_format=mf).eval()
    rnd = Random(42)
    w = h = 1024
    print(f'MPx\tGPU Ms {torch.__version__} {net.__class__.__name__} {dtype} {mf}')
    with torch.no_grad():
        for _ in range(1000):
            px = (rnd.random() * 3.9 + 0.1) * 1e6
            aspect_ratio = rnd.random() + 0.5
            w = int(math.sqrt(px * aspect_ratio))
            h = int(px / w)
            x = torch.empty((1, 4, w, h), dtype=dtype, device='cuda', memory_format=mf)
            t0 = time.time()
            y = net(x)
            torch.cuda.synchronize()
            dt = time.time() - t0
            print(f'{w * h * 1e-6:.3f}\t{dt * 1000:.1f}')
```
When I run this on my Win10+3090 machine I get: 
![image](https://github.com/pytorch/pytorch/assets/90846/09a0734c-4eaa-4dc5-b5f3-0eebde99563a)
I.e. significant difference in the scatter. There are also two significant outliers for 2.0.1 (not shown in the graph) of 33 seconds and 14 seconds. 

Even disregarding the outliers, 2.0.1 averages 2x the time 1.10.0 takes.

**Comment by jxtps** [comment]
A more complex network shows more complex behavior: 
```python
import math
import os
import time
from random import Random

os.environ['PYTORCH_CUDA_ALLOC_CONF'] = 'max_split_size_mb:512'
os.environ['TORCH_CUDNN_V8_API_DISABLED'] = '1'

import torch

# Download u2net implementation from:
# https://raw.githubusercontent.com/xuebinqin/U-2-Net/master/model/u2net_refactor.py
from u2net_refactor import U2NET_full

if __name__ == '__main__':
    dtype = torch.float16
    use_channels_last = False
    mf = torch.channels_last if use_channels_last else torch.contiguous_format
    net = U2NET_full().to(dtype=dtype, device='cuda', memory_format=mf).eval()
    rnd = Random(42)
    w = h = 1024
    print(f'MPx\tGPU Ms {torch.__version__} {net.__class__.__name__} {dtype} {mf}')
    with torch.no_grad():
        for _ in range(1000):
            px = (rnd.random() * 1.9 + 0.1) * 1e6
            aspect_ratio = rnd.random() + 0.5
            w = int(math.sqrt(px * aspect_ratio))
            h = int(px / w)
            x = torch.empty((1, 3, w, h), dtype=dtype, device='cuda', memory_format=mf)
            t0 = time.time()
            y = net(x)
            torch.cuda.synchronize()
            dt = time.time() - t0
            print(f'{w * h * 1e-6:.3f}\t{dt * 1000:.1f}')
```
When I run this on my Win10+3090 machine I get:
![image](https://github.com/pytorch/pytorch/assets/90846/6730cd08-7bb3-4de1-b6d3-68dd7dfaf97c)
"Faceted" scatter with 2.0.1 & 1.12.1, and 1.10.0 significantly outperforms. Averages: 
1. 1.10.0: 68.4ms
2. 1.12.1: 117.6ms
3. 2.0.1: 114.7ms

**Comment by ptrblck** [comment]
Not reproducible for me on a 3090 by removing the env variables and adding the missing sync using `2.1.0.dev20230817+cu121` on Linux.

**Comment by jxtps** [comment]
This is really strange. Installed the latest nightly, `2.2.0.dev20230906+cu121` & commented out the env-vars. 

When I run the simple network, I get *amazing* performance: 
![image](https://github.com/pytorch/pytorch/assets/90846/5a33f1e8-e2ca-4831-b4cc-25b970fb15b3)

But when I run the more complex network I get *terrible* performance: 
![image](https://github.com/pytorch/pytorch/assets/90846/aa3115f9-09fb-4d43-9ddf-69c732106a58)

???

**Comment by jxtps** [comment]
Does this match what you see @ptrblck ?

**Comment by eqy** [comment]
Mentioning a workaround if you do not care about host-side memory usage that I commented in the PR: Have you checked if `TORCH_CUDNN_V8_LRU_CACHE_LIMIT=0` changes performance to be more favorable in your case?

**Comment by jxtps** [comment]
After the lack of response I kind of gave up on this. Using variable sized inputs was really convenient, but seems poorly supported not just from a performance perspective, but also from a stability perspective: when I switched to tile & stitch the process stopped crashing every so often (though that may have been due to other factors, not sure).

**Comment by eqy** [comment]
`TORCH_CUDNN_V8_API_DISABLED=1` uses the cuDNN legacy API and is effectively maintenance-only. If you can articulate what constitutes terrible performance (e.g., `TORCH_CUDNN_V8_API_LRU_CACHE_LIMIT=0` still produces slowdowns over time with dynamic shapes) or have reproducible crashes presumably caused by cuDNN, then we would be interested in fixing these.

**Comment by jxtps** [comment]
I commented out the env-vars for the latest run with the mixed performance: 

```python
import math
import os
import time
from random import Random

# os.environ['PYTORCH_CUDA_ALLOC_CONF'] = 'max_split_size_mb:512'
# os.environ['TORCH_CUDNN_V8_API_DISABLED'] = '1'

import torch

# Download u2net implementation from:
# https://raw.githubusercontent.com/xuebinqin/U-2-Net/master/model/u2net_refactor.py
from u2net_refactor import U2NET_full

if __name__ == '__main__':
    dtype = torch.float16
    use_channels_last = False
    mf = torch.channels_last if use_channels_last else torch.contiguous_format
    net = U2NET_full().to(dtype=dtype, device='cuda', memory_format=mf).eval()
    rnd = Random(42)
    w = h = 1024
    print(f'MPx\tGPU Ms {torch.__version__} {net.__class__.__name__} {dtype} {mf}')
    with torch.no_grad():
        for _ in range(1000):
            px = (rnd.random() * 1.9 + 0.1) * 1e6
            aspect_ratio = rnd.random() + 0.5
            w = int(math.sqrt(px * aspect_ratio))
            h = int(px / w)
            x = torch.empty((1, 3, w, h), dtype=dtype, device='cuda', memory_format=mf)
            t0 = time.time()
            y = net(x)
            torch.cuda.synchronize()
            dt = time.time() - t0
            print(f'{w * h * 1e-6:.3f}\t{dt * 1000:.1f}')
```

**Comment by Ghost-LZW** [comment]
I did some perf on this and fixed them in this PR https://github.com/pytorch/pytorch/pull/114110. I'll show my perf result in this PR, you can help me review this if you are interested. @jxtps @eqy

**Comment by jxtps** [comment]
Reran this with Pytorch 2.5.1, and the simple TestModel network still gets amazing performance: 

![Image](https://github.com/user-attachments/assets/e949fafb-c9ec-40b7-a264-65a8989fa15d)

And now the more complex U2NET also gets consistent performance (even tough pytorch 1.10.0+cu113 is still beating it): 

![Image](https://github.com/user-attachments/assets/84f84b1b-fe30-4674-a5f2-bd3940a36156)

I really appreciate the consistency achievements, but the 50ms 0.5Mpx latency with 1.10.0 was pretty phenomenal - any chance of clawing that back?

**Comment by eqy** [comment]
This looks like it could be CPU overhead, what happens if you disable the V8 API now?

**Comment by jxtps** [comment]
It gets slightly worse on the low MPx end, incrementally better at 2 MPx, and has a wider scatter: 

![Image](https://github.com/user-attachments/assets/d733e26c-a239-4410-954e-ff257eccce02)

**Comment by jxtps** [comment]
Oops, that was with channels_last, here is the contiguous format version: 

![Image](https://github.com/user-attachments/assets/0f70ef66-af42-4e59-9357-9fe4f1c55e70)

It's basically worse across the board.
